enhancements/monitoring/integration-with-topology-modes.md

```diff
@@ -0,0 +1,182 @@
+as one of the Prometheus instances is still up.
+
+The Thanos ruler instances of user workload monitoring work the same as
+Prometheus except they only deal with rule evaluations.
+
+The 3 Alertmanager instances receive alerts from both Prometheus instances and
+as long as one instance is up and running, the notifications should flow to the
+receivers. Alertmanager instances also replicate silences and notification logs
+between themselves to prevent duplicate notifications.
+
+The other monitoring components running with multiple replicas in HA mode
+(thanos-querier and prometheus-adapter) are stateless. Running multiple
+replicas only ensures that their services remain available as long as 1
+instance is up.
+
+By definition, the monitoring stack wouldn't be highly-available in the
+`SingleReplica` mode. One recommendation would be that cluster admins configure
+Prometheus and Alertmanager with persistent storage to ensure that existing
+metrics, silences and notification logs are retained upon pod recreation.
+
+## Design Details
+
+### Test Plan
+
+We will setup a CI job that runs the `extended` end-to-end test suite from
+[OpenShift Origin](https://github.com/openshift/origin) in a single-node
+deployment mode. The test suite has already been extended to verify that
+operators honor the `SingleReplica` value (see
+[openshift/origin#25812](https://github.com/openshift/origin/pull/25812) and
+[openshift/origin/pull/25885](https://github.com/openshift/origin/pull/25885)).
+
+As soon as the implementation in CMO is merged, we will adjust the single-node
+topology
+[tests](https://github.com/openshift/origin/blob/master/test/extended/single_node/topology.go)
+to remove the exception in place for the monitoring components.
+
+Cluster admins can use the `openshift-monitoring/cluster-monitoring-operator`
+configmap to tune the configuration of the operator and its operands (
+user-workload monitoring enablement, persistent storage, resource
+requests/limits, ...). We have [end-to-end tests](https://github.com/openshift/cluster-monitoring-operator/tree/master/test/e2e)
+in the cluster-monitoring-operator verifying that these configuration
+parameters work as expected. Because the CMO configuration doesn't allow to
+change the replica count and it doesn't expose the topology mode either, we
+don't plan to add any test specific to the single-node mode to the CMO test
+suite.
+
+### Graduation Criteria
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+As stated in the [Single-node Production Deployment Approach](https://github.com/openshift/enhancements/blob/master/enhancements/single-node/production-deployment-approach.md)
+proposal, in-place upgrades are a non-goal right now.
+
+### Version Skew Strategy
+
+If CMO reads a value that is neither equal to `HighAvailable` nor
+`SingleReplica`, it should consider that the topology is `HighAvailable`.
+
+If CMO gets an error while reading the resource, it errors out and retries.
```

Comment:
what decision does it make about the number of replicas to configure in this case, though?

Reply:
I'd say:
* if it gets a "not found" error, it assumes `HighAvailable`.
* if it gets any other error, it should do nothing and retry reading the resource. 
What do you think?

Reply:
seems reasonable to me i think.

Reply:
Do we have functional configurations in which the Infrastructure config resource isn't present at all? That didn't come up in the enhancement describing the API https://github.com/openshift/enhancements/blob/master/enhancements/single-node/cluster-high-availability-mode-api.md

Reply:
Agree that the `infrastructure/cluster` resource should always be there but I don't think it hurts to cover this case. @dhellmann do you think that CMO should bail out when the resource isn't found?

Reply:
Assuming a default value for the API seems reasonable, but we should be consistent with other operators. What do other operators do when they can't find that particular resource?

Reply:
I've searched within the openshift org and the only thing that pops up is the cluster ingress operator [1] which bails out on any failure to retrieve the cluster resource.

[1] https://github.com/openshift/cluster-ingress-operator/blob/a5525f387106974730ccb9f422e05115498cbc47/pkg/operator/operator.go#L179-L182

Reply:
OK, that would put the pod into a crash-loop, until the resource is created. That seems like a reasonable approach, too.

CMO only runs on OpenShift, right? You don't need to worry about valid environments without that CR on other distributions?

Reply:
surely we can do better than a crashloop....

Reply:
Well, not "crash loop" I guess. But the reconcile loop would error and retry, right? That seems like the correct behavior.

Reply:
yes, retry (preferably on a delay, not in a hotloop) on error is reasonable.

Reply:
Makes sense, I've updated the text accordingly.